Implement fibers for win32

straight-shoota opened the pull request:
This PR continues the efforts to bring concurrency features to windows (see #5430).

There are some refactorings to the the existing Fiber, Thread and EventLoop classes to extract the platform-specific implementation focused on POSIX and libevent.

Then it adds implementations of these classes for windows. `Thread` and `EventLoop` are only stubbed so far.
The core functionality is the context swap on win32.

`concurrent_spec` passes and channels work in general, but the channel spec seems to hang.

Supersedes #6955

cfsamson commented on src/fiber.cr:
```diff
@@ -46,9 +45,21 @@ class Fiber
+      # It's the caller's responsibility to allocate 32 bytes of "shadow space" on the stack right
+      # before calling the function (regardless of the actual number of parameters used)
+
+      stack_ptr = @stack_bottom - sizeof(Void*) * 4
```
I have been thinking a bit about the shadow space, since at first I didn't think that we needed it at all. I have concluded that we do need it but I think the way we do it here gives the wrong end result. 

Right now we first reserve 4 bytes of shadow space, and then write `rsp` over one of them when writing `stack_ptr[0] = fiber_main.pointer`. I think we have to do like I show in this diagram since the next 16 byte aligned offset is 48 bytes from the bottom of the stack.
<img width="431" alt="Skjermbilde 2019-07-26 kl  23 25 14" src="https://user-images.githubusercontent.com/8337848/61986188-bdfc4200-b00d-11e9-957c-0f791c998026.png">
(it's just exemplified with "addresses" on a 64 byte stack).

The fact that we don't have a return address should not matter, the important part is that `rsp` ends up in the shadow space which might be used when debugging for example. `stack_ptr = @stack_bottom - sizeof(Void*) * 6` should avoid that.

Take a look at how [Boost Context does it](https://github.com/boostorg/context/blob/d4608a4e8ba2e8a0bdd45613243e6c53a23d23dc/src/asm/ontop_x86_64_ms_pe_gas.asm#L190). I linked to the place int the asm where they set up`rsp` and compare that offset with the position in the chart in the top of the file.
 
@ysbaddaden had some questions about the shadow space earlier and i have some references here besides the Boost Context one:
https://docs.microsoft.com/en-us/cpp/build/stack-usage?view=vs-2019#stack-allocation
https://github.com/simon-whitehead/assembly-fun/tree/master/windows-x64#shadow-space
https://wiki.lazarus.freepascal.org/Win64/AMD64_API#Shadow_space

straight-shoota commented:
@ysbaddaden in https://github.com/crystal-lang/crystal/pull/6955#issuecomment-515912947 (answered here in the new PR):

> About the rework: I meant that POSIX and Win32 don't need to share a base Thread class. The amount of duplicated code isn't large enough to justify the split. That would allow them to have their own internal details, so we treat them as "opaque classes".

Sure, we can do that. I don't know yet what the proper `Thread` implementation for win32 will look like, so I just did the minimal required thing to get fibers working.

We need to consider: Once we get proper multithreading, I suppose `Thread` might become a public API. I'm not sure if that's planned yet, but at least I think it would make sense to allow access to threads directly. In that case, we'll need to define a single public interface, like we do with other system-specific types.

> I'd also like this rework to settle on where we should put specific classes, namely Thread, Thread::Mutex, Thread::ConditionVariable (pthread, win32), EventLoop (libevent, win32, then probably epoll, kqueue, ...) to avoid further conflicts.
> 
> Being system specific, I suppose they should be moved under src/crystal/system?
> 
> For threads, I guess they should be unix/pthread.cr, unix/pthread_mutex.cr, win32/thread.cr, and so on, then have src/thread.cr, src/thread/mutex.cr require the correct one?

Sounds good.

> For the event loop, maybe unix/libevent.cr and win32/iocp.cr then have src/crystal/event_loop.cr require the correct one?

I'd like to include `event_loop` in the system-specific file names. So `unix/event_loop_libevent.cr` and `win32/event_loop_iocp.cr`.
For now, I've moved them all to `src/crystal/event_loop` because it's not exactly system-specific. Basically, the `libevent` bindings for unix should in theory work with win32 as well. But yeah, it probably still makes more sense to put that in `src/crystal/system/{unix,win32}`.

> Just a question: would it be hard to bring real win32 thread support, instead of bringing a stub? I know that's not needed to implement fibers, but threads aren't needed on POSIX either, and it would be awesome to have a basic support for them.

I really don't know. Haven't looked into that yet. And I'd rather like to get fibers working right away because that's the limiting factor for porting other features as well.

RX14 commented on src/thread.cr:
```diff
@@ -1,16 +1,19 @@
-require "./thread/*"
+require "./thread/linked_list"
+{% if flag?(:win32) %}
+  {% raise "thread not supported" %} 
```
This feels like it should be `Crystal::System`...

RX14 commented on src/crystal/event_loop.cr:
```diff
@@ -1,60 +1,30 @@
 end
+
+{% if flag?(:unix) %}
+  require "./event/event_loop_libevent"
```
This feels like it should be in `Crystal::System` too, but less so.

RX14 commented on src/lib_c/x86_64-windows-msvc/c/winbase.cr:
```diff
@@ -90,4 +90,16 @@ lib LibC
+  MEM_RESET      = 0x00080000
+  MEM_RESET_UNDO = 0x01000000
+
+  fun VirtualAlloc(lpAddress : Void*, dwSize : SizeT, flAllocationType : DWORD, flProtect : DWORD) : Void*
```
These two functions are actually in `memoryapi.h`, according to MSDN, not winbase

RX14 commented on src/crystal/scheduler.cr:
```diff
@@ -97,11 +97,23 @@ class Crystal::Scheduler
     exit 1
   end
 
+  private def print_error(message)
```
Maybe move this to `Crystal::System.print_error`?

RX14 commented on src/fiber/context/x86_64.cr:
```diff
@@ -3,39 +3,78 @@
-      pushq %r15
-      movq %rsp, 0($0)  // current_context.stack_top = %rsp
-      movl $$1, 8($0)   // current_context.resumable = 1
+    {% if flag?(:win32) %}
```
Should we have a `x86_64-windows-msvc.cr` and a `x86_64-linux-gnu.cr`?

straight-shoota replied:
Sounds reasonable.

ysbaddaden replied:
Why `x86_64-linux-gnu.cr`? The context switch is valid for all System-V targets (linux gnu/musl, openbsd, freebsd, ...). 

Are there alternatives to win32-msvc ABI, for example is MinGW different? In that case probably extract `x86_64-windows-msvc.cr` and have a`x86_64-posix.cr`.

straight-shoota replied:
`x86_64-posix.cr` => `x86_64-unix.cr`? We're typically using `unix` everywhere else...

ysbaddaden replied:
Oh yeah, sure.

RX14 replied:
I didn't know the standard abi had a name! If it's called systemv, then it should be named that!

`x86_64-sysv.cr` and `x86_64-microsoft.cr` would be my preference

straight-shoota replied:
That naming looks good, see https://en.wikipedia.org/wiki/X86_calling_conventions#x86-64_calling_conventions

straight-shoota commented:
I pushed some updates but didn't move any files yet. I'm waiting for feedback on this before pushing things around.

ysbaddaden commented:
@straight-shoota `Thread` is meant to stay has a private internal API, there is no plan to eventually make it public.

straight-shoota commented:
I've moved `Thread` and `EventLoop` to `crystal/system`, split the `x86_64` fiber context  implementation and also squashed all those fixup commits.

Sija commented on src/crystal/system/print_error.cr:
```diff
@@ -0,0 +1,12 @@
+  # Prints directly to stderr without going through an IO.
+  # This is useful for error messages from components that are required for
+  # IO to work (fibers, scheduler, event_loop).
+  def self.print_error(message)
```
I'd add a simple logic for appending `\n` to the *message*, since it's the most common usage case.
Perhaps sth along the lines of `message = message + "\n" unless message.ends_with?('\n')`?

straight-shoota replied:
This is a very low-level method. I think it is fine without convenient, but ultimately unnecessary logic enhancements.

ysbaddaden commented on src/prelude.cr:
```diff
@@ -79,7 +79,7 @@ require "static_array"
 require "symbol"
 no_win require "system"
-no_win require "thread"
+require "thread"
```
Shouldn't it become `crystal/thread.cr`?

straight-shoota commented:
Yes, exactly. It kind of makes sense to have `Thread` in the `Crystal` namespace, but that breaks some things. That's the reason why the tests keep failing because there are ever more places with broken references. It's probably worth it, though. When it's not intended to be public API, then it should go into the `Crystal` namespace like everything else.

The PR changes everything to `Crystal::Thread` but that's easy to swing back should we decide to keep it in the top level namespace.

RX14 commented on src/windows_stubs.cr:
```diff
@@ -73,3 +73,11 @@ end
   LibC.Sleep(time.total_milliseconds)
 end
+
+class Crystal::Thread
```
Perhaps move this to `src/crystal/system/win32/thread.cr`

straight-shoota replied:
IIRC you previously asked to have this stub in `windows_stubs` :D

RX14 replied:
haha, well, oops

I think that was when everything was ifdefs?

Either way, I think i've changed my mind since the rest of `Thread` is a stub too. What do you think? It's not particularly important either way.

ysbaddaden replied:
I agree, let's prepare for the win32 implementations by having stubs defined at their expected place, with a TODO: implement :-)

straight-shoota commented on fiber.cr:
```diff
@@ -0,0 +1,22 @@
+puts "running"
```
Definitely :+1:

straight-shoota commented on src/crystal/system/event_loop.cr:
```diff
@@ -0,0 +1,32 @@
+module Crystal::EventLoop
```
I figured the `Crystal::System` namespace is only really useful when there are both a system-specific type and a general type in `Crystal`. Otherwise it only adds clutter.
There is however the possibility of name clashes with the compiler which also uses the `Crystal` namespace. But that's a general issue unless we put *everything* in `Crystal::System`.

RX14 replied:
I put everything which has a platform-specific implementation - and is using the new `Crystal::System` structure for it in `Crystal::System`.

I'm fine with not renaming, but I personally would have done it, since the files are in `src/crystal/system`.

straight-shoota replied:
I see it that way that the things in `Crystal::System` are not used anywhere directly except as an implementation detail of their specific counterpart in the public API. `EventLoop`, `Thread` etc. however are general APIs for Crystal's runtime. They only have system-specific implementations, but they're directly used in the runtime. `Crystal::EventLoop` for example is referenced 10 times in the stdlib (without specs), `Crystal::Thread` 17 times.

straight-shoota commented:
Rebased and squashed.

RX14 commented:
and CI failed

ysbaddaden commented:
> Unhandled exception: BUG: Thread.current returned NULL (Exception)

It fails to build crystal itself (ECR.embed for the init command).

straight-shoota commented:
@ysbaddaden `check_format` fails while building the compiler, but that's the only thing being built in that spec. All `test` checks fail while executing `ecr/process` in a `run` macro for `ecr_spec`. `ecr/process` is probably just the first crystal program executed with the changed (local) standard library-

To reproduce the error directly:
```console
$ bin/crystal run src/ecr/process.cr -- spec/std/data/test_template2.ecr
Unhandled exception: BUG: Crystal::Thread.current returned NULL (Exception)
  from src/raise.cr:191:1 in 'raise'
  from src/raise.cr:229:3 in 'raise'
  from src/crystal/system/unix/pthread.cr:64:3 in 'current'
  from src/crystal/scheduler.cr:17:5 in 'enqueue'
  from src/concurrent.cr:63:3 in 'spawn'
  from src/kernel.cr:540:1 in '__crystal_main'
  from src/crystal/main.cr:97:5 in 'main_user_code'
  from src/crystal/main.cr:86:7 in 'main'
  from src/crystal/main.cr:106:3 in 'main'
  from __libc_start_main
  from _start
  from ???
```

Out of a hunch, I re-inserted `require "thread"` into [scheduler.cr](https://github.com/crystal-lang/crystal/pull/7995/commits/f47ec7f6bfdc2d646f0ba109230baf277dd547d0#diff-3920767d4904eb70109bd332570d8cc4L3) (which seemed unnecessary at first), but it makes the error vanish. I guess there is some effect from require order. CI should pass now.

Sija commented on src/crystal/system/win32/thread.cr:
```diff
@@ -0,0 +1,74 @@
+
+  # Returns the Thread object associated to the running system thread.
+  def self.current : Thread
+    @@current || raise "BUG: Thread.current returned NULL"
```
```suggestion
    @@current || raise "BUG: Crystal::Thread.current returned NULL"
```

straight-shoota commented:
Rebased on master with #8032 to see test_linux32 succeed.

asterite commented on src/fiber.cr:
```diff
@@ -46,9 +45,21 @@ class Fiber
-    # point to first addressable pointer on the stack (@stack_bottom points past
-    # the stack because the stack grows down):
-    stack_ptr = @stack_bottom - sizeof(Void*)
+    # FIXME: This line shouldn't be necessary (#7975)
```
One way to remove this FIXME is to do the if flag check twice, but the second one inside the assignment to stack_ptr. But it's not necessary, I think this way is fine too.

asterite commented:
Once we merge this, how far would we be regarding running Crystal in Windows?

wontruefree commented:
@asterite looks like there are eight more bullet points to work on.
https://github.com/crystal-lang/crystal/issues/5430

asterite commented:
This now needs to be rebased on top of `master`.

asterite commented on src/crystal/system/thread_linked_list.cr:
```diff
@@ -1,13 +1,13 @@
 
 # :nodoc:
-class Thread
+class Crystal::Thread
```
Could we revert this change and leave `Thread` instead of `Crystal::Thread`? I don't think we can merge this PR with this. But @waj and @bcardiff might be able to provide more context about why we don't want to make this change.

RX14 replied:
It should at least be another PR, honestly.

I like the change though.

asterite replied:
Maybe. I just don't think we should delay Windows support just because some want to rename a class and some don't.

straight-shoota replied:
Okay, I'll revert to `Thread`.

ysbaddaden commented on src/crystal/system/win32/event_loop_iocp.cr:
```diff
@@ -0,0 +1,70 @@
+module Crystal::EventLoop
+  @@queue = Deque(Fiber).new
+
+  private def self.loop_fiber
```
The EventLoop API changed a lot with MT. There is no more loop fiber, and less methods.

straight-shoota commented:
I've updated the win32 implementation to reflect the recent changes to `EventLoop` from #8173.
Also rebased on top of #8195 to get it compile on win32.

RX14 commented:
Another coflict?

RX14 commented:
would love another review

firejox commented on src/fiber/context/x86_64-microsoft.cr:
```diff
@@ -0,0 +1,78 @@
+          popq %rbx
+          popq %rdi         // pop 1st argument (for initial resume)
+          popq %gs:0x08
+          popq %gs:0x10
```
The order it popped is different from the order it stored in `makecontext`. Is it correct?

ysbaddaden replied:
Sorry, I didn't mean to resolve the discussion....

I believe it's a documentation error of setting rbx in makecontext (or we push rbx when we don't need to?).

We only pop up till rcx and the return address is left on the stack, I think.

firejox replied:
@ysbaddaden  Well, I meant these `popq %gs:0x08` and `popq %gs:0x10` two instructions , not about `rbx` and `rcx`. Sorry for including too large region. 

In `makecontext`, they store in the following order.
```
    # The following two values are stored in the Thread Information Block (NT_TIB)
    # and are used by Windows to track the current stack limits
    stack_ptr[-2] = @stack_bottom # %gs:0x08: Stack Bottom
    stack_ptr[-3] = @stack        # %gs:0x10: Stack Limit
```

Shouldn't the pop order be this?
```
  popq %gs:0x10
  popq %gs:0x08
```

cfsamson replied:
I think you're right. `GS:[0x08]` is the "high address", and `GS:[0x10]` is the "low address" and if I'm not mistaken `@stack_bottom` is the high address, and `@stack` is the low address so they're put on the initial stack in the wrong order (or we need to change both push and pop orders in `asm`).

I would just change the initial setup to:
```
stack_ptr[-2] = @stack              # %gs:0x10: Stack Limit
stack_ptr[-3] = @stack_bottom # %gs:0x08: Stack Bottom
```
@ysbaddaden might be right that we don't need to push/pop `%rcx` on every switch. I don't have the setup here to confirm that, but it might be a few cycles wasted there.

ysbaddaden replied:
Ah yes. They're inverted in makecontext. I totally missed that.
